spec.html

```diff
@@ -27849,123 +28115,85 @@ <h1>
             _specifier_: a String,
-            _promiseCapability_: a PromiseCapability Record,
-            _innerPromise_: a Promise,
+            _state_: a ModuleLoadState Record,
+            _result_: either a normal completion containing a Module Record or a throw completion,
           ): ~unused~
         </h1>
         <dl class="header">
           <dt>description</dt>
-          <dd>FinishDynamicImport completes the process of a dynamic import originally started by an <emu-xref href="#sec-import-calls">`import()`</emu-xref> call, resolving or rejecting the promise returned by that call as appropriate according to _innerPromise_'s resolution. It is performed by host environments as part of HostImportModuleDynamically.</dd>
+          <dd></dd>
         </dl>
         <emu-alg>
-          1. Let _fulfilledClosure_ be a new Abstract Closure with parameters (_result_) that captures _referencingScriptOrModule_, _specifier_, and _promiseCapability_ and performs the following steps when called:
-            1. Assert: _result_ is *undefined*.
-            1. Let _moduleRecord_ be ! HostResolveImportedModule(_referencingScriptOrModule_, _specifier_).
-            1. Assert: Evaluate has already been invoked on _moduleRecord_ and successfully completed.
-            1. Let _namespace_ be Completion(GetModuleNamespace(_moduleRecord_)).
-            1. If _namespace_ is an abrupt completion, then
-              1. Perform ! Call(_promiseCapability_.[[Reject]], *undefined*, &laquo; _namespace_.[[Value]] &raquo;).
-            1. Else,
-              1. Perform ! Call(_promiseCapability_.[[Resolve]], *undefined*, &laquo; _namespace_.[[Value]] &raquo;).
-            1. Return ~unused~.
-          1. Let _onFulfilled_ be CreateBuiltinFunction(_fulfilledClosure_, 0, *""*, &laquo; &raquo;).
-          1. Let _rejectedClosure_ be a new Abstract Closure with parameters (_error_) that captures _promiseCapability_ and performs the following steps when called:
-            1. Perform ! Call(_promiseCapability_.[[Reject]], *undefined*, &laquo; _error_ &raquo;).
-            1. Return ~unused~.
-          1. Let _onRejected_ be CreateBuiltinFunction(_rejectedClosure_, 0, *""*, &laquo; &raquo;).
-          1. Perform PerformPromiseThen(_innerPromise_, _onFulfilled_, _onRejected_).
-          1. Return ~unused~.
+          1. If _result_ is a normal completion, then
+            1. If _referrer_.[[LoadedModules]] contains a Record _record_ such that _record_.[[Specifier]] is _specifier_, then
+              1. Assert: _record_.[[Module]] is _result_.[[Value]].
+            1. Else, add Record { [[Specifier]]: _specifier_, [[Module]]: _result_.[[Value]] } to _referrer_.[[LoadedModules]].
+          1. If _state_.[[Action]] is ~graph-loading~, then
+            1. Perform ContinueModuleLoading(_state_, _result_).
+          1. Else,
+            1. Perform ContinueDynamicImport(_state_, _result_).
```

Comment:
Why do dynamic imports not go through recursive `LoadRequestedModules` calls the same as static imports? Would such work be possible here or as a follow-up, or does it involve non-trivial integration changes to the dynamic import hooks?

Reply:
They do, it's at step 8 of `ContinueDynamicImport`.

Reply:
That's amazing you could work the full unification here, thanks.